PhysicsTools/PatAlgos/plugins/PackedCandidateTrackChi2Producer.cc

```diff
@@ -0,0 +1,113 @@
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "DataFormats/Common/interface/ValueMap.h"
+#include "DataFormats/PatCandidates/interface/Muon.h"
+#include "DataFormats/TrackReco/interface/Track.h"
+#include "DataFormats/PatCandidates/interface/PackedCandidate.h"
+#include "DataFormats/Math/interface/libminifloat.h"
+
+namespace pat {
+
+  class PackedCandidateTrackChi2Producer : public edm::stream::EDProducer<> {
+    typedef edm::ValueMap<float> FloatMap;
+
+  public:
+    explicit PackedCandidateTrackChi2Producer(const edm::ParameterSet& iConfig)
+        : candidateToken_(consumes<pat::PackedCandidateCollection>(iConfig.getParameter<edm::InputTag>("candidates"))),
+          trackToken_(consumes<reco::TrackCollection>(iConfig.getParameter<edm::InputTag>("trackCollection"))),
+          doLostTracks_(iConfig.getParameter<bool>("doLostTracks")) {
+      if (doLostTracks_) {
+        track2LostTrackToken_ = consumes<edm::Association<pat::PackedCandidateCollection>>(
+            iConfig.getParameter<edm::InputTag>("candidates"));
+      } else {
+        candidate2PFToken_ =
+            consumes<edm::Association<reco::PFCandidateCollection>>(iConfig.getParameter<edm::InputTag>("candidates"));
+      }
+
+      produces<FloatMap>();
+    }
+
+    void produce(edm::Event&, const edm::EventSetup&) override;
+
+    static void fillDescriptions(edm::ConfigurationDescriptions&);
+
+  private:
+    const edm::EDGetTokenT<pat::PackedCandidateCollection> candidateToken_;
+    edm::EDGetTokenT<edm::Association<reco::PFCandidateCollection>> candidate2PFToken_;
+    edm::EDGetTokenT<edm::Association<pat::PackedCandidateCollection>> track2LostTrackToken_;
+    const edm::EDGetTokenT<reco::TrackCollection> trackToken_;
+    const bool doLostTracks_;
+  };
+
+}  // namespace pat
+
+void pat::PackedCandidateTrackChi2Producer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  edm::Handle<pat::PackedCandidateCollection> candidates;
+  iEvent.getByToken(candidateToken_, candidates);
+
+  const edm::Association<reco::PFCandidateCollection>* candidate2PF = nullptr;
+  if (!doLostTracks_) {
+    candidate2PF = &iEvent.get(candidate2PFToken_);
+  }
+
+  const edm::Association<pat::PackedCandidateCollection>* tracks2LT = nullptr;
+  edm::Handle<reco::TrackCollection> trks;
+  if (doLostTracks_) {
+    tracks2LT = &iEvent.get(track2LostTrackToken_);
+    iEvent.getByToken(trackToken_, trks);
+  }
+
+  const auto nCand = candidates->size();
+  std::vector<float> trkChi2Map(nCand, 0);
+
+  if (doLostTracks_) {  //for Lost tracks we don't have references to PFCands, so we must loop over tracks and check keys...
+    for (size_t i = 0; i < trks->size(); i++) {
+      float nChi2 = 0;
```

Comment:
remove the `float nChi2` allocation here and just move it directly to line 73, it is not used outside of the `doLostTracks_`.